exercises/react/example/Reactor.ceylon

```diff
@@ -0,0 +1,88 @@
+import ceylon.collection { HashMap, MutableMap }
+
+alias Element => Integer;
+
+class Reactor() {
+  shared abstract class Cell() {
+    shared formal Element currentValue;
+    variable ComputeCell[] dependents = [];
+    shared void addDependent(ComputeCell c) {
+      dependents = dependents.withTrailing(c);
+    }
+    shared void eachDependent(Anything(ComputeCell) f) {
+      for (dep in dependents) {
+        f(dep);
+      }
+    }
+  }
+
+  shared class InputCell(Element initialValue) extends Cell() {
+    variable Element currentValue_ = initialValue;
+    shared actual Element currentValue => currentValue_;
+    assign currentValue {
+      currentValue_ = currentValue;
+      eachDependent((d) => d.propagate());
+      eachDependent((d) => d.fireCallbacks());
+    }
+  }
+
+  shared InputCell newInputCell(Element initialValue) {
+    return InputCell(initialValue);
+  }
+
+  shared class ComputeCell(Element() newValue) extends Cell() {
+    shared alias Callback => Anything(Element);
+
+    variable Element currentValue_ = newValue();
+    variable Element lastCallbackValue = currentValue_;
+    shared actual Element currentValue => currentValue_;
+
+    variable Integer callbacksIssued = 0;
+    variable MutableMap<Integer, Callback> activeCallbacks = HashMap<Integer, Callback>();
+
+    shared interface Subscription {
+      shared formal void cancel();
+    }
+
+    shared Subscription addCallback(Callback f) {
+      value id = callbacksIssued;
+      callbacksIssued++;
+      activeCallbacks.put(id, f);
+      return object satisfies Subscription {
+        cancel() => activeCallbacks.remove(id);
+      };
+    }
+
+    shared void propagate() {
+      value nv = newValue();
+      if (nv != currentValue) {
+        currentValue_ = nv;
+        eachDependent((d) => d.propagate());
+      }
+    }
+
+    shared void fireCallbacks() {
+      if (lastCallbackValue == currentValue) {
+        return;
+      }
+      lastCallbackValue = currentValue;
+      for (cb in activeCallbacks.items) {
+        cb(currentValue);
+      }
+      eachDependent((d) => d.fireCallbacks());
+    }
+  }
+
+  shared ComputeCell newComputeCell1(Cell c, Element(Element) f) {
```

Comment:
https://github.com/exercism/xceylon/pull/20#discussion_r96332335 explains why `Reactor` must be in charge of creating new compute cells - notice the `addDependent` call on the newly-created cell. If this was done in an `ComputeCell` initialiser, it will leak a reference to `this`, which is illegal.